Lib/test/test_imp.py

```diff
@@ -318,6 +330,16 @@ def test_load_source(self):
         with self.assertRaisesRegex(ValueError, 'embedded null'):
             imp.load_source(__name__, __file__ + "\0")
 
+    @requires_create_dynamic
+    def test_issue31315(self):
+        # There shouldn't be an assertion failure in imp.create_dynamic(),
+        # when spec.name is not a string.
+        class BadSpec:
+            name = 42
+            origin = 'foo'
```

Comment:
If origin is not a string SystemError can be raised. This should be fixed too.

Reply:
what code causes a SystemError?
I got a `TypeError: bad argument type for built-in operation` both on my Windows and Ubuntu.

Reply:
Your are right. I thought `_PyUnicode_AsUnicode()` raises SystemError, but it raises TypeError. No additional check is needed.